packages/@react-aria/menu/src/useSubMenuTrigger.ts

```diff
@@ -0,0 +1,244 @@
+  node: RSNode<unknown>,
+  /** Whether the submenu trigger is disabled. */
+  isDisabled?: boolean,
+  // TODO: naming. Also talk about if this should be customizable in this hook or if it belongs somewhere else
+  /** Type of the submenu being rendered. */
+  subMenuType?: 'dialog' | 'menu',
+  /** Ref of the menu that contains the submenu trigger. */
+  parentMenuRef: RefObject<HTMLElement>,
+  /** Ref of the submenu opened by the submenu trigger. */
+  subMenuRef: RefObject<HTMLElement>
+}
+
+export interface SubMenuTriggerAria<T> {
+  subMenuTriggerProps: AriaMenuItemProps,
+  subMenuProps: AriaMenuOptions<T>,
+  popoverProps: Pick<AriaPopoverProps, 'isNonModal'>,
+  // TODO: descriptions
+  // TODO: perhaps cherry pick these props from existing types, comes from several different types though
+  overlayProps:
+  {
+    onExit: () => void,
+    disableFocusManagement: boolean,
+    shouldCloseOnInteractOutside: (element: Element) => boolean
+  }
+}
+
+// TODO description
+// TODO: debatable if we should have a useSubMenu hook for the submenu keyboard handlers. Feels better to have it here so we don't need to ferry around
+// things like parentMenuRef and the open timeout canceling since we centralized them here
+export function UNSTABLE_useSubMenuTrigger<T>(props: AriaSubMenuTriggerProps, state: SubMenuTriggerState, ref: RefObject<FocusableElement>): SubMenuTriggerAria<T> {
+  let {parentMenuRef, subMenuRef, subMenuType = 'menu', isDisabled, node} = props;
+  let subMenuTriggerId = useId();
+  let overlayId = useId();
+  let {direction} = useLocale();
+  let openTimeout = useRef<ReturnType<typeof setTimeout> | undefined>();
+  let cancelOpenTimeout = useCallback(() => {
+    if (openTimeout.current) {
+      clearTimeout(openTimeout.current);
+      openTimeout.current = undefined;
+    }
+  }, [openTimeout]);
+
+  let onSubmenuOpen = useEffectEvent((focusStrategy?: FocusStrategy) => {
+    cancelOpenTimeout();
+    state.open(focusStrategy);
+  });
+
+  let onSubMenuClose = useEffectEvent(() => {
+    cancelOpenTimeout();
+    state.close();
+  });
+
+  useLayoutEffect(() => {
+    return () => {
+      cancelOpenTimeout();
+    };
+  }, [cancelOpenTimeout]);
+
+
+  // TODO: problem with setting up the submenu close handlers here is that we don't get access to the onClose the user provides
+  // to the menu since this is too far up . Maybe just provide the subMenuTriggerState and have useMenu and useMenuItem handle the key handlers internally?
+  // Maybe we should just have the user pass something like onMenuClose to the trigger level since we have the same problem of being unable to call onClose if the user interacts outside?
+  // Maybe have Menu call onClose in an effect when it detects that it is unmounting? StrictMode is problematic since the cleanup fires even when the component isn't actually unmounting
+  // (not a problem if we wanna keep onClose to fire only if the user selects a item)
```

Comment:
Followup work captured in https://github.com/orgs/adobe/projects/19/views/20?pane=issue&itemId=39054342 for getting onClose to fire whenever a menu/submenu closes rather than only when the user selects an item. Ideally it could happen on Menu unmount but had some problems with StrictMode and effect cleanup behavior. Also would require useMenu to accept a new prop alongside `onClose` since we need `onClose` to still close all menus on item select but the user provided onClose should only fire in said effect cleanup or whatever method of calling it on unmount we go with